pkg/filestore/s3/s3.go

```diff
@@ -0,0 +1,198 @@
+// limitations under the License.
+
+package s3
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/filestore"
+)
+
+type Store struct {
+	client          *s3.S3
+	bucket          string
+	profile         string
+	credentialsFile string
+
+	logger *zap.Logger
+}
+
+type Option func(*Store)
+
+func WithLogger(logger *zap.Logger) Option {
+	return func(s *Store) {
+		s.logger = logger.Named("s3")
+	}
+}
+
+func WithCredentialsFile(path, profile string) Option {
+	return func(s *Store) {
+		s.profile = profile
+		s.credentialsFile = path
+	}
+}
+
+func NewStore(region, bucket string, opts ...Option) (*Store, error) {
+	if region == "" {
+		return nil, fmt.Errorf("region is required field")
+	}
+	if bucket == "" {
+		return nil, fmt.Errorf("bucket is required field")
+	}
+
+	s := &Store{
+		bucket: bucket,
+		logger: zap.NewNop(),
+	}
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	sess, err := session.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a session: %w", err)
+	}
+	creds := credentials.NewChainCredentials(
+		[]credentials.Provider{
+			&credentials.EnvProvider{},
+			&credentials.SharedCredentialsProvider{
+				Filename: s.credentialsFile,
+				Profile:  s.profile,
+			},
+			&ec2rolecreds.EC2RoleProvider{
+				Client: ec2metadata.New(sess),
+			},
+		},
+	)
+	cfg := aws.NewConfig().WithRegion(region).WithCredentials(creds)
+	s.client = s3.New(sess, cfg)
+
+	return s, nil
+}
+
+func (s *Store) NewReader(ctx context.Context, path string) (rc io.ReadCloser, err error) {
+	input := &s3.GetObjectInput{
+		Bucket: aws.String(s.bucket),
+		Key:    aws.String(path),
+	}
+	out, err := s.client.GetObjectWithContext(ctx, input)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeNoSuchKey:
+				err = filestore.ErrNotFound
+				return
+			case s3.ErrCodeInvalidObjectState:
+				err = fmt.Errorf("invalid object state: %w", err)
+				return
+			default:
+				err = fmt.Errorf("unexpected aws error given: %w", err)
+				return
+			}
+		}
+		err = fmt.Errorf("unknown error given: %w", err)
+		return
+	}
+	return out.Body, nil
+}
+
+func (s *Store) GetObject(ctx context.Context, path string) (object filestore.Object, err error) {
+	rc, err := s.NewReader(ctx, path)
+	if err != nil {
+		return
+	}
+
+	content, err := ioutil.ReadAll(rc)
+	defer func() {
+		if e := rc.Close(); e != nil {
+			if err == nil {
+				err = fmt.Errorf("unable to close read object stream: %w", e)
+			} else {
+				err = fmt.Errorf("%w and unable to close read object stream: %s", err, e.Error())
+			}
+		}
```

Comment:
I'd say upon it calls `s.NewReader(ctx, path)`, should define this part. Besides, I think we need to check only error emitted by `rc.Close()` because `err` is being handled appropriately.

Reply:
That exactly what I want your point 👀 I don't have a strong opinion about this, but is it okay to just log the error given by `rc.Close()` with `s.logger` 👀

Reply:
Whether ignoring an error from Close() is good or bad. I think this topic is highly depending on its context, but in this case, logging could be enough. That get ends with success.

Reply:
But I also have a non-strong opinion. I'd be happy to hear your thoughts.

Reply:
Currently, in case of minio storage, we do ignore the handling (just show it on the log)
https://github.com/pipe-cd/pipe/blob/965abdc59242c57a5051ac3817d2bd583a5a28d4/pkg/filestore/minio/minio.go#L125
but in case of GCS we return it for further handling (maybe it's just be shown to the log somewhere either 😂 )
https://github.com/pipe-cd/pipe/blob/9ce7102afd38da0046e595f4dacc61aec9562e76/pkg/filestore/gcs/gcs.go#L108-L111
In short, we do not have a unified way to handle error from `rc.Close()` currently 😅 I think I will go as your recommended way to handle (logging with `s.logger` is enough). Let's me make separated PR to make the current GCS handling `rc.Close()` error to unify with other storages.

Reply:
Fine, I feel exactly the same way!

Reply:
okay let's me fix this part of s3 storage first 🤝

Reply:
Addressed on https://github.com/pipe-cd/pipe/pull/1485/commits/69df4dcbd640c91d45d6313cdce0e8cb82f29db4 🙏